fastcrypto/src/groups/bls12381.rs

```diff
@@ -333,6 +335,71 @@ impl Debug for G1Element {
 serialize_deserialize_with_to_from_byte_array!(G1Element);
 generate_bytes_representation!(G1Element, G1_ELEMENT_BYTE_LENGTH, G1ElementAsBytes);
 
+/// An uncompressed seralization of a G1 element. This format is two times longer than the compressed
+/// format used by `G1Element::serialize`, but is much faster to deserialize.
+///
+/// The intended use of this struct is to deserialize and sum a large number of G1 elements without
+/// having to decompress them first.
+#[derive(Clone, Debug)]
+#[repr(transparent)]
+pub struct G1ElementUncompressed(pub(crate) [u8; 2 * G1_ELEMENT_BYTE_LENGTH]);
+
+impl From<&G1Element> for G1ElementUncompressed {
+    fn from(element: &G1Element) -> Self {
+        let mut bytes = [0u8; 2 * G1_ELEMENT_BYTE_LENGTH];
+        unsafe {
+            blst_p1_serialize(bytes.as_mut_ptr(), &element.0);
+        }
+        G1ElementUncompressed(bytes)
+    }
+}
+
+impl G1ElementUncompressed {
+    /// Create a new `G1ElementUncompressed` from a byte array.
+    /// The input is not validated so it should come from a trusted source.
+    ///
+    /// See [the blst docs](https://github.com/supranational/blst/tree/master?tab=readme-ov-file#serialization-format) for details about the uncompressed serialization format.
+    pub fn from_trusted_byte_array(bytes: [u8; 2 * G1_ELEMENT_BYTE_LENGTH]) -> Self {
+        Self(bytes)
+    }
+
+    /// Get the byte array representation of this element.
+    pub fn into_byte_array(self) -> [u8; 2 * G1_ELEMENT_BYTE_LENGTH] {
+        self.0
+    }
+
+    fn to_blst_p1_affine(&self) -> FastCryptoResult<blst_p1_affine> {
+        let mut affine = blst_p1_affine::default();
+        unsafe {
+            // This fails if the point is not on the curve or if it is (0, ±2) which is on the curve
+            // but not in the G1 subgroup. See https://github.com/supranational/blst/blob/6f3136ffb636974166a93f2f25436854fe8d10ff/src/e1.c#L296-L326.
+            // A subgroup check is not performed here.
+            if blst_p1_deserialize(&mut affine, self.0.as_ptr()) != BLST_ERROR::BLST_SUCCESS {
+                return Err(InvalidInput);
+            }
+        }
+        Ok(affine)
+    }
+
+    /// Compute the sum of a slice of uncompressed G1 elements.
+    pub fn sum(terms: &[G1ElementUncompressed]) -> FastCryptoResult<G1Element> {
+        if terms.is_empty() {
+            return Ok(G1Element::zero());
+        }
+
+        let affine_points: Vec<blst_p1_affine> = terms
+            .iter()
+            .map(G1ElementUncompressed::to_blst_p1_affine)
+            .collect::<FastCryptoResult<Vec<_>>>()?;
+
+        // Inspired by https://github.com/supranational/blst/blob/6f3136ffb636974166a93f2f25436854fe8d10ff/bindings/rust/src/pippenger.rs#L334-L337
+        let mut ret = blst_p1::default();
+        let p: [*const _; 2] = [&affine_points[0], ptr::null()];
+        unsafe { blst_p1s_add(&mut ret, &p[0], terms.len()) };
```

Comment:
can you explain why the previous line is needed? why here it is not a simple pointer?

Reply:
I'm not sure. This is how it's done in the Rust bindings in blst, and using `&affine_points[0].as_ptr()` instead gives indeterministic results, probably because it's reading into unallocated memory. I'll say we use this method and test it throughly.

Reply:
let's add a link to the c++ code